src/server/session.rs

```diff
@@ -92,7 +92,231 @@ where
         self
     }
 
-    // Lots of code duplication here. Should disappear completely when this module in rewritten in async/.await style
+    /// Processing for the data connection. This will spawn a new async task with the actual processing.
+    ///
+    /// socket: the data socket we'll be working with
+    /// tls: tells if this should be a TLS connection
+    /// tx: channel to send the result of our operation to the control process
+    pub(super) fn spawn_data_processing(&mut self, socket: tokio02::net::TcpStream, tx: Sender<InternalMsg>) {
+        let mut data_cmd_rx = self.data_cmd_rx.take().unwrap().fuse();
+        let mut data_abort_rx = self.data_abort_rx.take().unwrap().fuse();
+        let tls = self.data_tls;
+        let mut command_executor = DataCommandExecutor {
+            user: self.user.clone(),
+            socket,
+            tls,
+            tx,
+            cmd: None,
+            storage: Arc::clone(&self.storage),
+            cwd: self.cwd.clone(),
+            start_pos: self.start_pos,
+            identity_file: if tls { Some(self.certs_file.clone().unwrap()) } else { None },
+            identity_password: if tls { Some(self.certs_password.clone().unwrap()) } else { None },
+        };
+
+        tokio02::spawn(async move {
+            #[allow(unused_assignments)]
```

Comment:
Please extract the
```suggestion
            match incoming {
                Some(DataCommand::Abort) => {
                    info!("Abort received");
                }
                Some(DataCommand::ExternalCommand(command)) => {
                    info!("Data command received");
                    command_executor.cmd = Some(command);
                    command_executor.execute().await;
                }
                None => {
                    // This probably happened because the control channel was closed before we got here
                    warn!("Nothing received");
                }
            }
```
into a function and then there is no need for the
```
#[allow(unused_assignments)]
let mut incoming = None;
```
the extracted function can be called within the select!

Reply:
Cool idea, will try

Reply:
See if that works for you.